### ISSUE TITLE
NameError: name 'CriterionType' is not defined

### ISSUE THREAD

**Issue by benam2** [flashlight/flashlight#1025]
This is the error I get:

```
NameError                                 Traceback (most recent call last)
/tmp/ipykernel_18106/1529090364.py in <module>
     31 
     32 upstream = 'wav2vec2_hug_base_960'
---> 33 runner = Runner(args, config)

~/Desktop/ASR/s3prl/s3prl/downstream/runner.py in __init__(self, args, config)
     50         self.upstream = self._get_upstream()
     51         self.featurizer = self._get_featurizer()
---> 52         self.downstream = self._get_downstream()
     53         self.all_entries = [self.upstream, self.featurizer, self.downstream]
     54 

~/Desktop/ASR/s3prl/s3prl/downstream/runner.py in _get_downstream(self)
    131             upstream_rate = self.featurizer.model.downsample_rate,
    132             **self.config,
--> 133             **vars(self.args)
    134         ).to(self.args.device)
    135 

~/Desktop/ASR/s3prl/s3prl/downstream/
```

I reviewed this issue but I did not help me https://github.com/flashlight/flashlight/issues/416#issue-783305697

1. when I run this `import flashlight`  --> does not raises any error
2. when I run this `from flashlight.lib.text.decoder import CriterionType`   --> it raises this: `ModuleNotFoundError: No module named 'flashlight.lib.text'`
3. I can successfully run examples in the flashlight without any error. so this command `python flashlight/bindings/python/example/criterion_example.py` does NOT raise any error
4. And I have `kenlm` in this address `/usr/local/share/kenlm`

@jacobkahn I tried to address the questions @tlikhomanenko had asked in that issue in order to make it easier.

I really really appreciate it if you could give me some hints where should I check.

**Comment by benam2** [comment]
I really appreciate it if you can have a quick look at this issue, I've included all the information and steps I took and was mentioned in other issue pages to fix it but no luck :(. @jacobkahn @tlikhomanenko 
Thanks again for your time

**Comment by benam2** [comment]
I don't know if it can help but this are the warnings I get before raising error:

```
/home/ai-labs/.conda/envs/asr/lib/python3.7/site-packages/torchaudio/_internal/module_utils.py:87: UserWarning: Failed to import soundfile. 'soundfile' backend is not available.
  warnings.warn("Failed to import soundfile. 'soundfile' backend is not available.")
################################################################################
### WARNING, path does not exist: KALDI_ROOT=/mnt/matylda5/iveselyk/Tools/kaldi-trunk
###          (please add 'export KALDI_ROOT=<your_path>' in your $HOME/.profile)
###          (or run as: KALDI_ROOT=<your_path> python <your_script>.py)
################################################################################

[hubconf] can not import .upstream.pase.hubconf: No module named 'pase'... Pass.
[Override] - config.downstream_expert.datarc.decoder_args.decoder_type = kenlm
[Override] - config.downstream_expert.datarc.decoder_args.kenlm_model = ../../backup/4-gram.arpa.gz
[Override] - config.downstream_expert.datarc.decoder_args.lexicon = ../../backup/librispeech_lexicon.lst
/home/ai-labs-sgoudarzvand/.conda/envs/asr/lib/python3.7/site-packages/transformers/configuration_utils.py:365: UserWarning: Passing `gradient_checkpointing` to a config initialization is deprecated and will be removed in v5 Transformers. Using `model.gradient_checkpointing_enable()` instead, or if you are using the `Trainer` API, pass `gradient_checkpointing=True` in your `TrainingArguments`.
  "Passing `gradient_checkpointing` to a config initialization is deprecated and will be removed in v5 "
Some weights of the model checkpoint at facebook/wav2vec2-base were not used when initializing Wav2Vec2Model: ['quantizer.weight_proj.bias', 'project_q.bias', 'project_q.weight', 'project_hid.bias', 'quantizer.weight_proj.weight', 'quantizer.codevectors', 'project_hid.weight']
- This IS expected if you are initializing Wav2Vec2Model from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing Wav2Vec2Model from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
[Featurizer for <class 's3prl.upstream.wav2vec2_hug.expert.UpstreamExpert'>] - The input upstream is only for initialization and not saved in this nn.Module
[Featurizer for <class 's3prl.upstream.wav2vec2_hug.expert.UpstreamExpert'>] - Take a list of 13 features and weighted sum them.
[Runner] - Loading Featurizer weights from the previous experiment
/home/ai-labs/Desktop/ASR/s3prl/s3prl/downstream/asr/w2l_decoder.py:42: UserWarning: flashlight python bindings are required to use this functionality. Please install from https://github.com/facebookresearch/flashlight/tree/master/bindings/python
  "flashlight python bindings are required to use this functionality. Please install from [https://github.com/facebookresearch/flashlight/tree/master/bindings/python"](https://github.com/facebookresearch/flashlight/tree/master/bindings/python%22)
```

**Comment by benam2** [comment]
Any help on this please @jacobkahn @tlikhomanenko

**Comment by jacobkahn** [comment]
@benam2 opened https://github.com/s3prl/s3prl/issues/407 -- I'll play around with this for a minute to see if there's an easy fix, but the bindings haven't changed.

**Comment by benam2** [comment]
I really appreciate it. I have already installed `flashlight text` as you suggested but did not fix this error. Does that mean for some reason it has not been installed correctly? (didn't raise any error though)

This is the last part of the log in terminal

```
copying build/lib.linux-x86_64-3.7/flashlight/lib/text/flashlight_lib_text_dictionary.cpython-37m-x86_64-linux-gnu.so -> build/bdist.linux-x86_64/egg/flashlight/lib/text
byte-compiling build/bdist.linux-x86_64/egg/flashlight/lib/__init__.py to __init__.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/flashlight/lib/text/dictionary.py to dictionary.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/flashlight/lib/text/decoder.py to decoder.cpython-37.pyc
creating build/bdist.linux-x86_64/egg/EGG-INFO
copying flashlight_text.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO
copying flashlight_text.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying flashlight_text.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying flashlight_text.egg-info/not-zip-safe -> build/bdist.linux-x86_64/egg/EGG-INFO
copying flashlight_text.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
writing build/bdist.linux-x86_64/egg/EGG-INFO/native_libs.txt
creating dist
creating 'dist/flashlight_text-0.1-py3.7-linux-x86_64.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing flashlight_text-0.1-py3.7-linux-x86_64.egg
removing '/home/ai-labs-sgoudarzvand/.conda/envs/asr/lib/python3.7/site-packages/flashlight_text-0.1-py3.7-linux-x86_64.egg' (and everything under it)
creating /home/ai-labs-sgoudarzvand/.conda/envs/asr/lib/python3.7/site-packages/flashlight_text-0.1-py3.7-linux-x86_64.egg
Extracting flashlight_text-0.1-py3.7-linux-x86_64.egg to /home/ai-labs-sgoudarzvand/.conda/envs/asr/lib/python3.7/site-packages
flashlight-text 0.1 is already the active version in easy-install.pth

Installed /home/ai-labs-sgoudarzvand/.conda/envs/asr/lib/python3.7/site-packages/flashlight_text-0.1-py3.7-linux-x86_64.egg
Processing dependencies for flashlight-text==0.1
Finished processing dependencies for flashlight-text==0.1
```
@jacobkahn

**Comment by benam2** [comment]
I did not uninstall the flashlight, could it be the reason? @jacobkahn

**Comment by jacobkahn** [comment]
@benam2 — did you install Flashlight Text -- https://github.com/flashlight/text?
```
from flashlight.lib.text.decoder import CriterionType
```
works for me on a fresh install of Flashlight Text

**Comment by benam2** [comment]
@jacobkahn Yes I did as you had suggested in another issue. I followed this:

```
git clone https://github.com/flashlight/text && cd text
cd bindings/python
python3 setup.py install
```
and this is what I see in condo env: `flashlight-text           0.1                      pypi_0    pypi`

**Comment by jacobkahn** [comment]
@benam2 

And
```
from flashlight.lib.text.decoder import CriterionType
```
still fails? Can you check that you're using the correct Python interpreter in all cases?

**Comment by benam2** [comment]
Im trying to run `https://github.com/s3prl/s3prl/blob/main/s3prl/downstream/asr/w2l_decoder.py` script and it raises error while reaches the line to import `from flashlight.lib.text.decoder import CriterionType`.

Yea I have activated the env for the `asr` and then try to get that script running. I will create a new env again and try from scratch. Thanks for taking the time. I will share If I was successful.

**Comment by benam2** [comment]
@jacobkahn Just before starting everything from scratch, could you please comment on this error log that anything else looks suspicious to you. I really appreciate it if you can have a quick look:

```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Input In [3], in <cell line: 33>()
     30 hack_isinstance()
     32 upstream = 'wav2vec2_hug_base_960'
---> 33 runner = Runner(args, config)

File ~/Desktop/ASR/s3prl/s3prl/downstream/runner.py:52, in Runner.__init__(self, args, config)
     50 self.upstream = self._get_upstream()
     51 self.featurizer = self._get_featurizer()
---> 52 self.downstream = self._get_downstream()
     53 self.all_entries = [self.upstream, self.featurizer, self.downstream]

File ~/Desktop/ASR/s3prl/s3prl/downstream/runner.py:129, in Runner._get_downstream(self)
    127 def _get_downstream(self):
    128     Downstream = getattr(downstream, self.args.downstream)
--> 129     model = Downstream(
    130         upstream_dim = self.featurizer.model.output_dim,
    131         upstream_rate = self.featurizer.model.downsample_rate,
    132         **self.config,
    133         **vars(self.args)
    134     ).to(self.args.device)
    136     return self._init_model(
    137         model = model,
    138         name = 'Downstream',
    139         trainable = True,
    140         interfaces = ['get_dataloader', 'log_records']
    141     )

File ~/Desktop/ASR/s3prl/s3prl/downstream/asr/expert.py:99, in DownstreamExpert.__init__(self, upstream_dim, upstream_rate, downstream_expert, expdir, **kwargs)
     94 self.objective = nn.CTCLoss(
     95     blank=self.blank,
     96     zero_infinity=self.datarc['zero_infinity']
     97 )
     98 decoder_args = self.datarc.get('decoder_args')
---> 99 self.decoder = get_decoder(decoder_args, self.dictionary)
    100 self.register_buffer('best_score', torch.ones(1) * 100)

File ~/Desktop/ASR/s3prl/s3prl/downstream/asr/expert.py:36, in get_decoder(decoder_args_dict, dictionary)
     34     if isinstance(decoder_args.unk_weight, str):
     35         decoder_args.unk_weight = eval(decoder_args.unk_weight)
---> 36     return W2lKenLMDecoder(decoder_args, dictionary)
     38 return None

File ~/Desktop/ASR/s3prl/s3prl/downstream/asr/w2l_decoder.py:126, in W2lKenLMDecoder.__init__(self, args, tgt_dict)
    125 def __init__(self, args, tgt_dict):
--> 126     super().__init__(args, tgt_dict)
    128     self.unit_lm = getattr(args, "unit_lm", False)
    130     if args.lexicon:

File ~/Desktop/ASR/s3prl/s3prl/downstream/asr/w2l_decoder.py:55, in W2lDecoder.__init__(self, args, tgt_dict)
     52 self.nbest = args.nbest
     54 # criterion-specific init
---> 55 self.criterion_type = CriterionType.CTC
     56 self.blank = (
     57     tgt_dict.index("<ctc_blank>")
     58     if "<ctc_blank>" in tgt_dict.indices
     59     else tgt_dict.bos()
     60 )
     61 if "<sep>" in tgt_dict.indices:

NameError: name 'CriterionType' is not defined

```
@jacobkahn

**Comment by jacobkahn** [comment]
@benam2 — nothing interesting happening here besides that Text isn't properly installed -- your goal should be to get
```
from flashlight.lib.text.decoder import CriterionType
```
working

**Comment by benam2** [comment]
By chance I just tried to just import the library and not run the code and it does not raise any error.
It seems that it has been installed correctly since `from flashlight.lib.text.decoder import CriterionType` does not raise any error :|

Do you still think that the error log I shared before looks ok? @jacobkahn

**Comment by jacobkahn** [comment]
@benam2 if the import succeeds, it means Flashlight Text is installed correctly and that there's an issue with s3prl. You should open an issue in their repo.

**Comment by benam2** [comment]
Im really sorry to get back to you again. I did test a lot of things. The problem is not with s3prl script I think.
I created a new env, and then checked the code.

In the script where it imports the library it had used `try except` that's why it did not refer to the exact error line correctly. What I did, I just remove the `try/except`. it means I changed this lines:

```
try:
    from flashlight.lib.text.dictionary import create_word_dict, load_words
    from flashlight.lib.sequence.criterion import CpuViterbiPath, get_data_ptr_as_bytes
    from flashlight.lib.text.decoder import (
        CriterionType,
        LexiconDecoderOptions,
        KenLM,
        LM,
        LMState,
        SmearingMode,
        Trie,
        LexiconDecoder,
    )
except:
    warnings.warn(
        "flashlight python bindings are required to use this functionality. Please install from https://github.com/facebookresearch/flashlight/tree/master/bindings/python"
    )
```

to this line:

```
from flashlight.lib.text.dictionary import create_word_dict, load_words
from flashlight.lib.sequence.criterion import CpuViterbiPath, get_data_ptr_as_bytes
from flashlight.lib.text.decoder import CriterionType, LexiconDecoderOptions,KenLM, LM, LMState,SmearingMode,Trie,LexiconDecoder
```

Now the error is :

```
  File "run_downstream.py", line 206, in <module>
    main()
  File "run_downstream.py", line 201, in main
    runner = Runner(args, config)
  File "/home/ai-labs/Desktop/ASR/s3prl/s3prl/downstream/runner.py", line 52, in __init__
    self.downstream = self._get_downstream()
  File "/home/ai-lab/Desktop/ASR/s3prl/s3prl/downstream/runner.py", line 129, in _get_downstream
    model = Downstream(
  File "/home/ai-labs/Desktop/ASR/s3prl/s3prl/downstream/asr/expert.py", line 99, in __init__
    self.decoder = get_decoder(decoder_args, self.dictionary)
  File "/home/ai-labs/Desktop/ASR/s3prl/s3prl/downstream/asr/expert.py", line 32, in get_decoder
    from .w2l_decoder import W2lKenLMDecoder
  File "/home/ai-labs/Desktop/ASR/s3prl/s3prl/downstream/asr/w2l_decoder.py", line 29, in <module>
    from flashlight.lib.sequence.criterion import CpuViterbiPath, get_data_ptr_as_bytes
ModuleNotFoundError: No module named 'flashlight.lib.sequence'
```

And when I just do this:
`from flashlight.lib.sequence.criterion import CpuViterbiPath`

it raises:
`ModuleNotFoundError: No module named 'flashlight.lib.sequence'`


So it means even from the beginning it had problem with this with `flashlight.lib.sequence` (I can import `flashlight.lib without error).


Could you please comment on this that what could be incorrect?

**Comment by jacobkahn** [comment]
@benam2 you still to install sequence bindings which are still in this repo.

You were having trouble with FFTW before -- you can disable installing it with [this diff](https://gist.github.com/jacobkahn/f345a94974838c89df6b07f77824fca5).

End to end flow that works for me:
```bash
git clone https://github.com/flashlight/flashlight.git
cd flashlight
wget https://gist.githubusercontent.com/jacobkahn/f345a94974838c89df6b07f77824fca5/raw/65c3ca05961a92a08fbe1e2629a9290774baf6fd/diff.out
git apply diff.out
USE_CUDA=0 USE_MKL=0 pip install -e bindings/python
python -c "import flashlight.lib.sequence" # succeeds
```

I'm in the process of moving `flashlight.lib.sequence` bindings to another repo this week, but this should unblock you for the time being. Sorry for the build challenges here.

**Comment by jacobkahn** [comment]
@benam2 The package you're trying to use doesn't use ASG, which is the only thing in the sequence library that is eligible to be build with CUDA. Building with it won't affect performance since that codepath isn't being called but you can enable it if you feel like it :)

**Comment by benam2** [comment]
@jacobkahn 
So this is the message for the flashlight `Successfully installed flashlight-1.0.0`
And this is the error after importing `ModuleNotFoundError: No module named 'flashlight.lib.sequence'`

**Comment by jacobkahn** [comment]
@benam2 — well — turns out this is a bug in some envs!

Fix is in https://github.com/flashlight/flashlight/pull/1032. I'll let you know when those both land and we can retry.

**Comment by benam2** [comment]
@jacobkahn the error related to `flashlight.lib.sequence` fortunately got fixed but I think there has been a change in `flashlight/text` that now it raises error. Should I wait for a new commit or is there anything I need to do in my end?

(If that helps, I just tried this approach on an older env that I had installed the `flashlight-text` successfully and I could import successfully yesterday. But now when I activated that env and just followed the first part of the instruction to get `flashlight.lib.sequence` fixed, `import flashlight.lib.text` also raises error.
What Im trying to say is that there is a change that affected `"import flashlight.lib.text"`.)

**Comment by jacobkahn** [comment]
@benam2 — you need to `pip uninstall flashlight` and `pip uninstall flashlight-text` before installing them both again, from source, with up to date branches.

**Comment by benam2** [comment]
I had uninstalled `flashlight-text` but same thing happened. I will try to uninstall both of them and if it did not work create a new env again. I hope that fix it. I will update here. Thanks again.

**Comment by benam2** [comment]
Uninstalling did not help. I go to create a new env and do everything again. Just one quick thing that `git apply diff.out` raises error but it installs successfully. Is that ok? @jacobkahn

**Comment by benam2** [comment]
@jacobkahn I created a new env, and followed your instruction:

1. python -c "import flashlight.lib.sequence".  Runs successfully
2. python -c "import flashlight.lib.text". Raises Error

and like I said `git apply diff.out` raises error but I think that is fine since the original error is fixed.
I use `python= 3.8`

Did you do it in conda env?

**Comment by jacobkahn** [comment]
@benam2:

Updated diff for debugging purposes: https://gist.githubusercontent.com/jacobkahn/4b26e5146b70c5cfc18c429a8bc9cc52/raw/be75c21d27e1ebe4f7e407a1b32316a7fe0637ed/diff.out

Can you nuke and reclone your Flashlight Text repo, `pip uninstall flashlight-text`, and install again? I've confirmed that everything works with multiple envs.

**Comment by benam2** [comment]
@jacobkahn Still raises error. Im gonna share what exactly Im doing so that if you confirm the steps I will spend some time to try different things.

These are my step:

```
1. pip uninstall flashlight
2. pip uninstall flashlight-text
3. sudo rm -r flashlight (so both flashlight and text are removed)
4. git clone https://github.com/flashlight/flashlight.git && cd flashlight
5. wget https://gist.githubusercontent.com/jacobkahn/4b26e5146b70c5cfc18c429a8bc9cc52/raw/be75c21d27e1ebe4f7e407a1b32316a7fe0637ed/diff.out
6. git apply diff.out (without error)
7. USE_CUDA=0 USE_MKL=0 pip install -e bindings/python
8. python -c "import flashlight.lib.sequence" # succeeds
9. Inside flashlight folder ---> git clone https://github.com/flashlight/text && cd text
10. cd bindings/python
11. python3 setup.py install
12. python -c "import flashlight.lib.text" # ERROR
```

 I have 1 question as well.

1. I remember initially I was trying to install the flashlight, MKL .... so they are somewhere in my system. do you think that can affect what Im experiencing now? if so I may need to remove my account. (But the `flashlight.lib.text` was running ok yesterday that's why I think that could not be the case).

**Comment by jacobkahn** [comment]
@benam2 — this is now an error with the package you're using. The error's likely that the file you're trying to load doesn't exist, but I'd open an issue with s3prl and they'll help you debug. Glad everything is working on the Flashlight side.

**Comment by benam2** [comment]
Actually the file exist and that was the first thing I checked. 
 The code in the s3prl is very straightforward in the sense that it raises error in this line `self.lexicon = load_words(args.lexicon)` and this load_words is one functionality in `flashlight-text` not something that `s3prl` has implemented. Could you please share your reasoning why this is not a `flashlight-text` error?

**Comment by benam2** [comment]
Alright, @jacobkahn  I will check various ways to see where could be wrong. thanks again. and I greatly appreciate your quick reply.
